packages/nimble-components/src/table/specs/table-row-grouping-hld.md

```diff
@@ -0,0 +1,238 @@
+-   [Table README](./README.md)
+-   [Table Design Doc](https://xd.adobe.com/view/5b476816-dad1-4671-b20a-efe796631c72-0e14/screen/d389dc1e-da4f-4a63-957b-f8b3cc9591b4/specs/)
+-   [Prototype branch](https://github.com/ni/nimble/tree/table-row-grouping-prototype) ([Storybook](https://60e89457a987cf003efc0a5b-bwzrahvxqm.chromatic.com/?path=/story/table--table))
+
+## Implementation / Design
+
+![Row Grouping Example](./spec-images/row-groups.PNG)
+
+### Row Grouping Column API
+
+To support both the notion that a column type may want to allow grouping in some situations and not others (e.g. a numeric column displaying discrete values versus continuous values), as well as the possibility that certain columns should never be groupable, we will create a mixin to provide the necessary grouping APIs:
+
+```
+export function mixinGroupableColumnAPI<
+    TBase extends GroupableTableColumnConstructor
+>(base: TBase) {
+    abstract class GroupableColumn extends base {
+        public disableGrouping?: boolean;
+
+        public groupIndex?: number | null = null;
+
+        public isDisableGroupingChanged(): void {
+            this.internalIsGroupable = !this.disableGrouping;
+        }
+
+        public groupIndexChanged(): void {
+            this.internalGroupIndex = this.groupIndex;
+        }
+
+        /**
+         * The custom element tag to use for rendering group header values.
+         * Should derive from TableGroupHeaderView.
+         */
+        public abstract groupHeaderViewTag: string;
+    }
+    attr({ attribute: 'disable-grouping', mode: 'boolean' })(GroupableColumn.prototype, 'disableGrouping');
+    attr({ attribute: 'group-index', converter: nullableNumberConverter })(GroupableColumn.prototype, 'groupIndex');
+
+    return GroupableColumn;
+}
+```
+
+The `TableColumn` will add the `internalIsGroupable` and `internalGroupIndex` APIs to provide that state where it is needed in the rest of the `Table` implementation.
+
+This mixin can then be chained to other mixins in the following fashion:
+
+```
+class TableColumnTextFractionalWidth extends mixinFractionalWidthColumnAPI(TableColumnTextBase) {}
+
+export class TableColumnText extends mixinGroupableColumnAPI(TableColumnTextFractionalWidth) {}
+```
+
+### Rendering group header values
+
+Rendering group header values have similar concerns as rendering cell values, providing reason to adopt the pattern outlined in [this PR](https://github.com/ni/nimble/pull/1052) for columns to specify a custom element to use to render the value. As shown in the mixin snippet above, it will provide an abstract `groupHeaderViewTag` property, which will require groupable columns to specify the element to use to display the header value. This element will derive from an element of type `TableGroupRowHeaderView`:
+
+```
+export interface TableGroupRowHeaderValueState<TColumnConfig = unknown> {
+    groupHeaderValue: TableFieldValue;
```

Comment:
We should be able to define a type for this analogous to `TableCellRecord` (although this is just a value and not a record).

i.e. a column author based on how they define the operationalDataField type should be able to specify the type associated with this value (i.e. a column author should be able to specify they know the possible types they need to handle are number | undefined and not all types of TableFieldValue)